lib/src/Pos/Logic/Full.hs

```diff
@@ -0,0 +1,246 @@
+import           Pos.Txp.Network.Listeners (TxpMode)
+import qualified Pos.Txp.Network.Listeners as Txp (handleTxDo)
+import           Pos.Txp.Network.Types (TxMsgContents (..))
+import qualified Pos.Update.Logic.Local as Update (getLocalProposalNVotes, getLocalVote,
+                                                   isProposalNeeded, isVoteNeeded)
+import           Pos.Update.Mode (UpdateMode)
+import qualified Pos.Update.Network.Listeners as Update (handleProposal, handleVote)
+import           Pos.Util.Chrono (NewestFirst, OldestFirst, NE)
+
+
+
+type LogicWorkMode ctx m =
+    ( HasConfiguration
+    , HasBlockConfiguration
+    , WithLogger m
+    , MonadReader ctx m
+    , HasPrimaryKey ctx
+    , MonadMask m
+    , MonadBlockDBRead m
+    , MonadDBRead m
+    , MonadGState m
+    , MonadRecoveryInfo m
+    , MonadSlots ctx m
+    , HasLens SecurityParams ctx SecurityParams
+    , HasLens RecoveryHeaderTag ctx RecoveryHeader
+    , BlockWorkMode ctx m
+    , DlgListenerConstraint ctx m
+    , TxpMode ctx m
+    , UpdateMode ctx m
+    )
+
+-- | A stop-gap full logic layer based on the RealMode. It just uses the
+-- monadX constraints to do most of its work.
+logicLayerFull
+    :: forall ctx m x .
+       ( LogicWorkMode ctx m
+       )
+    => (JLTxR -> m ())
+    -> (LogicLayer m -> m x)
+    -> m x
+logicLayerFull jsonLogTx k =
+    bracket acquire release $ \_ -> do
+
+        -- Delivered monadically but in fact is constant (comes from a
+        -- reader context).
+        ourStakeholderId <- getOurStakeholderId
+        securityParams <- view (lensOf @SecurityParams)
+
+        let
+            getBlock :: HeaderHash -> m (Maybe Block)
+            getBlock = DB.getBlock
+
+            getTip :: m Block
+            getTip = DB.getTipBlock
+
+            getTipHeader :: m BlockHeader
+            getTipHeader = DB.getTipHeader
+
+            getAdoptedBVData :: m BlockVersionData
+            getAdoptedBVData = gsAdoptedBVData
+
+            recoveryInProgress :: m Bool
+            recoveryInProgress = Recovery.recoveryInProgress
+
+            getBlockHeader :: HeaderHash -> m (Maybe BlockHeader)
+            getBlockHeader = DB.getHeader
+
+            getBlockHeaders
+                :: NonEmpty HeaderHash
+                -> Maybe HeaderHash
+                -> m (Either GetBlockHeadersError (NewestFirst NE BlockHeader))
+            getBlockHeaders checkpoints start = do
+                result <- runExceptT (DB.getHeadersFromManyTo checkpoints start)
+                either (pure . Left . GetBlockHeadersError) (pure . Right) result
+
+            getBlockHeaders'
+                :: HeaderHash
+                -> HeaderHash
+                -> m (Either GetBlockHeadersError (OldestFirst NE HeaderHash))
+            getBlockHeaders' older newer = do
+                outcome <- DB.getHeadersRange Nothing older newer
+                case outcome of
+                    Left txt -> pure (Left (GetBlockHeadersError txt))
```

Comment:
`first GetBlockHeadersError <$> DB.getHeadersRange Nothing older newer`

Reply:
Is what I've written here a violation of some style guide?

Reply:
Obviously, not. But it's easier to read and also super easy to fix, so i don't see any reason not to. Up to you, of course.